clients/roscpp/src/libros/spinner.cpp

```
@@ -136,27 +208,21 @@ AsyncSpinnerImpl::~AsyncSpinnerImpl()
 
 bool AsyncSpinnerImpl::canStart()
 {
-  boost::recursive_mutex::scoped_try_lock spinlock(spinmutex);
-  return spinlock.owns_lock();
+  return true;
 }
 
 void AsyncSpinnerImpl::start()
 {
   boost::mutex::scoped_lock lock(mutex_);
 
   if (continue_)
-    return;
+    return; // already spinning
 
-  boost::recursive_mutex::scoped_try_lock spinlock(spinmutex);
-  if (!spinlock.owns_lock()) {
-    ROS_WARN("AsyncSpinnerImpl: Attempt to start() an AsyncSpinner failed "
-             "because another AsyncSpinner is already running. Note that the "
-             "other AsyncSpinner might not be using the same callback queue "
-             "as this AsyncSpinner, in which case no callbacks in your "
-             "callback queue will be serviced.");
-    return;
+  if (!spinner_monitor.add(callback_queue_, false))
+  {
+    ROS_FATAL_STREAM("AsyncSpinnerImpl: " << DEFAULT_ERROR_MESSAGE);
+    throw std::runtime_error("There is already a single-threaded spinner on this queue");
```

This is a notable change in behavior to escalate to a runtime_error where the old behavior is to not register the callback queue. We could consider making this change in an upcoming release, but we should not make this change in  an existing distro.

---

I did this on purpose: Indeed, the old behavior simply warned on the console, but otherwise silently continued without registering the callback queue. So, if the user / developer doesn't notice the warning in first place, he might assume everything is fine, but events are never processed on this queue!
The code out there either doesn't encounter this error/exception because AsyncSpinner was correctly used beforehand or - if it encountered the warning - it wasn't processing its events at all, which (hopefully) would have been triggered the developer to look for the error too. Hence, probably / hopefully such code was never released.

---

I agree with the arguments from both of you. But since Kinetic has already been released for a while any behavior change should be avoided (no matter how unlikely it is that code relies on it). Therefore the same conclusion as above:
* `ROS_FATAL_STREAM` -> `ROS_ERROR_STREAM`
* `throw` -> `return`.

---

I see. However, this is not as trivial as changing `throw` -> `return`.

Previous behavior was to accept multiple spinners on a queue as long as they were _started_ from the same thread. To retain this wrong behavior, I need to remember which thread initially started spinning on a particular queue and then allow other threads as well.

Hence, the main improvement remaining from this PR in Kinetic, is the ability to have multiple spinners on different queues.

---

I don't see why we need `initial_tid` and the condition check based on it. We only need `tid` to distinguish single threaded and multi threaded spinners. The same queue can only be handled by one single threaded spinner *or* any number of multi threaded spinners - no matter what the thread id of the multi threaded spinner is.

---

@dirk-thomas What you describe is the new and intended behavior and I fully agree. However, @tfoote asked to maintain the previous behavior and not throw. Previously, we could have the following situations:

1. `S1` (single-threaded spinner started in `thread 1`): will block `thread 1` until shutdown. Any further spinners in different threads were not allowed (with an error message).
2. `M1` (multi-threaded spinner started in `thread 1`): Further spinners started from _different_ threads were not allowed (with an error message).
3. `M1 ... M1 S1` (multi-threaded spinners started in `thread 1` and afterwards a single-threaded one started): This was accepted without any errors. But the new behavior is to reject `S1`!

Restrictions of case 1 + 2 are relaxed with this PR: Other spinners are allowed as long as the operate on a different queue. Thread doesn't matter.

The tricky part is case 3, which - although nonsense - was perfectly valid code before.
In order to maintain the old behavior, I need to remember, which thread the first M-spinner was started in (the `initial_tid`). If I wouldn't store the `initial_tid`, but allow S on any thread, this would relax unintended behavior even beyond previous behavior.

---

The "collision" in case 3 doesn't matter in my opinion. The fact that from which thread the multi-threaded spinners have been started from shouldn't be considered imo. The only collision to avoid is which queues they handle and to avoid that a single-threaded spinner handles the same queue as a multi-threaded spinner.

---

I fully agree. But it wasn't like this before!
To retain old behavior as requested by Tully, I need to introduce `initial_id`. Actually, case 3 might be a common use case: starting some AsyncSpinners before finally entering the `ros::spin` loop.

---

As long as both spinners don't spin on the same queue I think that is totally fine - independent from which thread the multi-threaded one was started from. That should be achieve but just removing the `initial_tid` again and its check, correct?

---

If spinners operate on different queues, there is no conflict at all and spinners will be started without problems (this is the basic improvement we gain with this PR).

However, I thought we discuss the case, where spinners want to operate on the _same_ queue. There might be code out there, which hits case 3. Removing `initial_id` and the corresponding check, but allowing to start the S* spinner from any thread after M1 is operating on the queue, will be a _weaker check_ than before: The old code before this PR, at least rejected spinning if S* and M1 were started from different threads. Of course, we would like to reject always, but Tully requested to not do so.